pkg/kstatus/watcher/event_funnel.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2022 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package watcher
+
+import (
+	"context"
+
+	"sigs.k8s.io/cli-utils/pkg/kstatus/polling/event"
+)
+
+// eventFunnel wraps a list of event channels and multiplexes them down to a
+// single event channel. New input channels can be added at runtime, and the
+// output channel will remain open until all input channels are closed.
+type eventFunnel struct {
+	// ctx closure triggers shutdown
+	ctx context.Context
+	// outCh is the funnel that consumes all events from input channels
+	outCh chan event.Event
+	// doneCh is closed after outCh is closed.
+	// This allows blocking until done without consuming events.
+	doneCh chan struct{}
+	// counterCh is used to track the number of open input channels.
+	counterCh chan int
+}
+
+func newEventFunnel(ctx context.Context) *eventFunnel {
+	funnel := &eventFunnel{
+		ctx:       ctx,
+		outCh:     make(chan event.Event),
+		doneCh:    make(chan struct{}),
+		counterCh: make(chan int),
+	}
+	// Wait until the context is done. Then wait until the counter reaches zero,
+	// indicating all input channels are closed. Then close the out and done
+	// channels to signal completion.
+	go func() {
+		// Nil channels never receive events, so we can use them in a select
+		// statement to block until both channels are nil, but to make them nil,
+		// we need local variables.
+		ctxDoneCh := ctx.Done()
+		counterCh := funnel.counterCh
+
+		// Count input channels that have been added and not closed.
+		inputs := 0
+		for {
+			select {
+			case delta, open := <-counterCh:
+				if !open {
+					counterCh = nil
+				} else {
+					inputs += delta
+					// Counter reached zero after context was closed.
+					if ctxDoneCh == nil && inputs <= 0 {
+						close(counterCh)
```

Comment:
It's risky to close a channel someone might send to as they will panic. With this change calling `AddInputChannel()` is not safe as it will panic even if context is done.

WDYT of this approach?
```go
func newEventFunnel(ctx context.Context) *eventFunnel {
	funnel := &eventFunnel{
		ctx:       ctx,
		outCh:     make(chan Event),
		doneCh:    make(chan struct{}),
		counterCh: make(chan int),
	}
	// Wait until the context is done. Then wait until the counter reaches zero,
	// indicating all input channels are closed. Then close the out and done
	// channels to signal completion.
	go func() {
		defer close(funnel.outCh)
		defer close(funnel.doneCh)
		ctxDoneCh := ctx.Done()

		// Count input channels that have been added and not closed.
		inputs := 0
		for {
			select {
			case delta := <-funnel.counterCh:
				inputs += delta
			case <-ctxDoneCh:
				// Stop waiting for context closure.
				ctxDoneCh = nil
			}
			if ctxDoneCh == nil && inputs <= 0 {
				// Context is closed and all input channels are closed.
				break
			}
		}
	}()
	return funnel
}
```

Reply:
> With this change calling AddInputChannel() is not safe as it will panic even if context is done.

With the current impl, if AddInputChannel is called after the context is closed, it won't try to send +1 on the counterCh:
```
select {
	case <-m.ctx.Done(): // skip, if context is closed
		return
	case m.counterCh <- 1: // increment counter
	}
```

It may look like they execute in parallel, but selects execute top down, only attempting subsequent cases if the previous cases block, the same as if statements operating left to right.

Your code is a little more elegant, except you forgot to close the counterCh.

Reply:
Updated to use your approach with a few tweaks.

Reply:
> It may look like they execute in parallel, but selects execute top down, only attempting subsequent cases if the previous cases block, the same as if statements operating left to right.

https://go.dev/ref/spec#Select_statements says:

> If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection.

Sometimes this crashes, sometimes succeeds https://go.dev/play/p/X62ydvz9HKj

> Your code is a little more elegant, except you forgot to close the counterCh.

That was the whole point :) Why do we need to close it? If we close it then we risk crashing blocked and future callers of `AddInputChannel()`.

Reply:
> If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection.

I stand corrected. 

> That was the whole point :) Why do we need to close it?

I always figured it was good practice to close channels for cleanup, but according to this, they don't need to be closed unless you're waiting for them to close. https://stackoverflow.com/a/36613932

Reply:
Alright, I removed the `close(counterCh)` and added a `EventFunnelClosedError` to return from `AddInputChannel` if the context is closed. The handling of this error should help prevent informers from being started after the context is already cancelled.